homeassistant/components/systemmonitor/sensor.py

```diff
@@ -134,6 +154,14 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
             else:
                 resource[CONF_ARG] = ""
 
+            # Check if argument is required. If yes, but none was provided, log an error.
+            if SENSOR_TYPES[resource[CONF_TYPE]][4] is True:
```

Comment:
Please make a custom validator function and use that in the voluptuous platform schema instead. See eg the mysensors integration for how to write custom validators.

https://github.com/home-assistant/core/blob/c68056d3ad76bd601b2b6e344738c5d52b89596f/homeassistant/components/mysensors/__init__.py#L40

https://github.com/home-assistant/core/blob/c68056d3ad76bd601b2b6e344738c5d52b89596f/homeassistant/components/mysensors/__init__.py#L100-L102

Reply:
But can such a validation function control whether an vol element should be required or optional? Or would I have to keep the "arg" as optional and via the validator throw an error for those sensor where it actually should be required (which cannot be enforced statically here via the schema)?

Reply:
Yes, keep it optional and raise `vol.Invalid` in the custom validator.

Reply:
@MartinHjelmare I cannot find a way to access inside the validator the value of another element from the schema. In order to now if "arg" is correct, I need access to the element "type".

Reply:
Yes, the validator should wrap the whole dict that contains both type and arg. So the value that the validator receives will be that dict.

Reply:
See the example I linked above. There the validator receives a list of dicts.

Reply:
The important hint was to move the validator one schema level up 😄👍 .

Reply:
@MartinHjelmare The drawback with the new approach is of course that now either all sensors work or they all fail, since the whole platform setup is stopped if an "arg" is missing.

Reply:
That's how our config validation works. The user can use the check config feature now to make sure the config is correct. It's more clear and we keep the code cleaner.